src/plugins/data_source/server/auth_registry/authentication_methods_registry.ts

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { deepFreeze } from '@osd/std';
+import { AuthenticationMethod } from '../../server/types';
+
+export type IAuthenticationMethodRegistery = Omit<
+  AuthenticationMethodRegistery,
+  'registerAuthenticationMethod'
+>;
+
+export class AuthenticationMethodRegistery {
+  private readonly authMethods = new Map<string, AuthenticationMethod>();
+  /**
+   * Register a authMethods with function to return credentials inside the registry.
+   * Authentication Method can only be registered once. subsequent calls with the same method name will throw an error.
+   */
+  public registerAuthenticationMethod(method: AuthenticationMethod) {
+    if (this.authMethods.has(method.name)) {
+      throw new Error(`Authentication method '${method.name}' is already registered`);
+    }
+    this.authMethods.set(method.name, deepFreeze(method) as AuthenticationMethod);
+  }
+
+  public getAllAuthenticationMethods() {
```

Comment:
we can consider to make these get function to return readonly values

Reply:
While adding authentication method in map, we do deepFreeze object and then store. 

[deepFreeze](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/packages/osd-std/README.md#deepfreeze) converts the return type to Readonly variant recursively.

so by default getAllAuthenticationMethods returns readonly values. I have added [UT](https://github.com/bandinib-amzn/OpenSearch-Dashboards/blob/main/src/plugins/data_source/server/auth_registry/authentication_methods_registry.test.ts#L97-L108) to validate that as well.